src/sentry/users/models/user.py

```diff
@@ -0,0 +1,605 @@
+        if not existing:
+            return do_write()
+
+        # Re-use the existing user if merging is enabled.
+        if flags.merge_users:
+            return (existing.pk, ImportKind.Existing)
+
+        # We already have a user with this `username`, but merging users has not been enabled. In
+        # this case, add a random suffix to the importing username.
+        lock = locks.get(f"user:username:{self.id}", duration=10, name="username")
+        with TimedRetryPolicy(10)(lock.acquire):
+            unique_db_instance(
+                self,
+                self.username,
+                max_length=MAX_USERNAME_LENGTH,
+                field_name="username",
+            )
+
+            # Perform the remainder of the write while we're still holding the lock.
+            return do_write()
+
+    @classmethod
+    def sanitize_relocation_json(
+        cls, json: Any, sanitizer: Sanitizer, model_name: NormalizedModelName | None = None
+    ) -> None:
+        model_name = get_model_name(cls) if model_name is None else model_name
+        super().sanitize_relocation_json(json, sanitizer, model_name)
+
+        sanitizer.set_string(json, SanitizableField(model_name, "username"))
+        sanitizer.set_string(json, SanitizableField(model_name, "session_nonce"))
+
+    @classmethod
+    def handle_async_deletion(
+        cls,
+        identifier: int,
+        region_name: str,
+        shard_identifier: int,
+        payload: Mapping[str, Any] | None,
+    ) -> None:
+        from sentry.hybridcloud.rpc.caching import region_caching_service
+        from sentry.users.services.user.service import get_many_by_id, get_user
+
+        region_caching_service.clear_key(key=get_user.key_from(identifier), region_name=region_name)
+        region_caching_service.clear_key(
+            key=get_many_by_id.key_from(identifier), region_name=region_name
+        )
+
+    def handle_async_replication(self, region_name: str, shard_identifier: int) -> None:
+        from sentry.hybridcloud.rpc.caching import region_caching_service
+        from sentry.users.services.user.service import get_many_by_id, get_user
+
+        region_caching_service.clear_key(key=get_user.key_from(self.id), region_name=region_name)
+        region_caching_service.clear_key(
+            key=get_many_by_id.key_from(self.id), region_name=region_name
+        )
+        organization_service.update_region_user(
+            user=RpcRegionUser(
+                id=self.id,
+                is_active=self.is_active,
+                email=self.email,
+            ),
+            region_name=region_name,
+        )
+
+
+# HACK(dcramer): last_login needs nullable for Django 1.8
+User._meta.get_field("last_login").null = True
+
+
+# When a user logs out, we want to always log them out of all
+# sessions and refresh their nonce.
+@receiver(user_logged_out, sender=User)
+def refresh_user_nonce(
+    sender: User | RpcUser, request: HttpRequest, user: User | None, **kwargs: Any
```

Comment:
unsure if request can be `None` too ?

Reply:
All of the current callers provide the request, *but* signals make no promises on providing all of their parameters. Annotating this as `HttpRequest | None` seems wise.